src/runtime/linux/auxvec.rs

```diff
@@ -0,0 +1,89 @@
+//! Reads /proc/self/auxv on Linux systems
+
+use std::prelude::v1::*;
+use std::slice;
+use std::mem;
+
+/// Simple abstraction for the ELF Auxiliary Vector
+///
+/// the elf.h provide the layout of the single entry as auxv_t.
+/// The desugared version is a usize tag followed by a union with
+/// the same storage size.
+///
+/// Cache only the HWCAP and HWCAP2 entries.
+#[derive(Debug)]
+pub struct AuxVec {
+    hwcap: Option<usize>,
+    hwcap2: Option<usize>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[allow(dead_code)]
+/// ELF Auxiliary vector entry types
+///
+/// The entry types are specified in  [linux/auxvec.h][auxvec_h].
+///
+/// [auxvec_h]: https://github.com/torvalds/linux/blob/master/include/uapi/linux/auxvec.h
+pub enum AT {
+    /// CPU Hardware capabilities, it is a bitfield.
+    HWCAP = 16,
+    /// CPU Hardware capabilities, additional bitfield.
+    HWCAP2 = 26,
+}
+
+impl AuxVec {
+    /// Reads the ELF Auxiliary Vector
+    ///
+    /// Try to read `/proc/self/auxv`.
+    // TODO: Make use of getauxval once it is available in a
+    // reliable way.
+    pub fn new() -> Result<Self, ::std::io::Error> {
+        use std::io::Read;
+        let mut file = ::std::fs::File::open("/proc/self/auxv")?;
```

Comment:
Apparently it is common on Android for this file to not be readable (permissions wise), but `getauxval()` is widely available because bionic (their libc) has implemented it for some years now.